doc/architectural/background-concepts.md

```diff
@@ -764,16 +764,378 @@ Next, try running minisat on the following formula:
+track their entire execution history. But programs with loops can take
+arbitrarily many steps, or even run into infinite loops, so we cannot describe
+their behavior using a finite propositional formula in the way we have
+done above.
+
+There are multiple approaches to deal with this problem, all with different
+trade-offs. CPROVER chooses bounded model checking as the underlying approach.
+The idea is that we only consider program
+executions that are, in some measure, ''shorter'' than a given bound. This bound
+then implies an upper bound on the size of the required formulas, which makes
+the problem tractable.
+
+To make this concrete, let's go back to the factorial example. We consider
+the function in the following form (in CPROVER, we actually use a `goto`-based
+IR, but the `while`-based version is a bit simpler to understand):
+```C
+unsigned long factorial(unsigned n) {
+  unsigned long fac = 1;
+  unsigned int i = 1;
+  while (i <= n) {
+    fac *= i;
+    i = i+1;
+  }
+  return fac;
+}
+```
+We set the following ''length bound'': The loops in the program are allowed to
+be executed at most three times; we will ignore all other executions. With this
+in mind, we observe that the following two classes of programs behave
+exactly the same:
+```C
+/* Execute this loop at most n+1 times */
+while (e) {
+  body;
+}
+```
+and
+```C
+if (e) {
+  body;
+  /* Execute this loop at most n times */
+  while (e) {
+    body;
+  }
+}
+```
+This transformation is known as ''loop unrolling'': We can always replace a loop
+by an `if` that checks if we should execute, a copy of the loop body and the
+loop statement.
+
+So, to reason about the `factorial` function, we unroll the loop three times
+and then replace the loop with a special `IGNORE` statement. We get:
+```C
+unsigned long factorial(unsigned n) {
+  unsigned long fac = 1;
+  unsigned int i = 1;
+  if (i <= n) {
+    fac *= i;
+    i = i+1;
+    if (i <= n) {
+      fac *= i;
+      i = i+1;
+      if (i <= n) {
+        fac *= i;
+        i = i+1;
+        if (i <= n) {
+          IGNORE;
+        }
+    }
+  }
+  return fac;
+}
+```
+Since this program is now in a simpler form without loops, we can use the
+techniques we learned above to transform it into a propositional formula.
+First, we transform into SSA (with code guard variables already included):
+```C
+unsigned long factorial(unsigned n) {
+  unsigned long fac.1 = 1;
+  unsigned int i.1 = 1;
+  bit C1 = i.1 <= n;
+  if (C1) {
+    fac.2 = fac.1 * i.1;
+    i.2 = i.1+1;
+    bit C2 = i.2 <= n;
+    if (C2) {
+      fac.3 = fac.2 * i.2;
+      i.3 = i.2+1;
+      bit C3 = i.3 <= n;
+      if (C3) {
+        fac.4 = fac.3 * i.3;
+        i.4 = i.3+1;
+        bit C4 = i.4 <= n;
+        if (C4) {
+          IGNORE;
+        }
+    }
+  }
+  return Phi(C1, Phi(C2, Phi(C3, fac.4, fac.3), fac.2), fac.1);
+}
+```
+We translate `IGNORE` into the formula **false** - this will later allow
+us to rule out all paths that reach this point.
```

Comment:
Could we add a few words about how ruling out some paths reflects on the correctness of the CBMC result? If there is a violating assertion after a (too long) while-loop do we say that no violating assertion was reached or that the verification was ended before completing?

Reply:
I have an example about that a few paragraphs below. Is that sufficient, or should I state something here already?

Reply:
Could you add a forward reference?

Reply:
Done.